### PR TITLE
509 new sqw for multifit

### PR DESCRIPTION
**Pull request by cmarooney-stfc**
Multifit implementation now uses new sqw objects.

The existing multifit tests in test_multifit have been extended to dnd objects. The reference solutions have been obtained by running multifit with dnd objects converted from the sqw files provided in test_multifit, using the old sqw/dnd implementations in the master branch.

These tests only cover d1d as the existing sqw multifit tests are also only for d1d. However the implemntations are all in the superclasses DnDBase or SQWDnDBase.

Addresses #509 

Abandoned - superseded by #638

### CODE REVIEW

File: _test/test_multifit/test_multifit_horace_dnd_2.m [hunk 1]
```diff
@@ -0,0 +1,90 @@
+classdef test_multifit_horace_dnd_2 < TestCaseWithSave
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_2.m:1]: can this be renamed more meaningfully?

File: _test/test_multifit/test_multifit_horace_dnd_2.m [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            this.w1data = d1d(sqw_w1data); 
+            this.w2data = d1d(sqw_w2data); 
+            this.win=[this.w1data,this.w2data];     % combine the two cuts into an array of sqw objects and fit
+            
+            % Save output, if requrested
+            %this.save();
+        end
+        
+        % ------------------------------------------------------------------------------------------------
+        function this=test_fit(this)
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_2.m:45]: test fit does what??

this test looks to verify that the fore- and background functions can be reversed

File: _test/test_multifit/test_multifit_horace_dnd_1.m [hunk 3]
```diff
@@ -0,0 +1,182 @@
+classdef test_multifit_horace_dnd_1 < TestCaseWithSave
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_1.m:1]: can this test be named more clearly

File: _test/test_multifit/test_multifit_horace_dnd_1.m [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            this.w1data = d1d(sqw_w1data);
+            this.w2data = d1d(sqw_w2data);
+            this.win=[this.w1data,this.w2data];     % combine the two cuts into an array of d1d objects and fit
+            
+            % Save output, if requrested
+            %this.save();
+        end
+        
+        % ------------------------------------------------------------------------------------------------
+        function this = test_fit_two_datasets(this)
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_1.m:49]: there's no test that it fits one dataset

File: _test/test_multifit/test_multifit_horace_dnd_1.m [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        
+        % ------------------------------------------------------------------------------------------------
+        function this=DISABLED_test_fit_array_of_datasets(this)
+            % Example of fitting single datasets, and independently fitting an array of datasets
+            % This exercises fit_sqw for single datasets and for an array of datasets
+
+            % *** NEEDS FIT_SQW TO BE IMPLEMENTED
+
+            % Check fit_sqw correctly fits array of input
+%             [wfit_single1,fitpar_single1]=fit_sqw(this.w1data, @sqw_bcc_hfm, [5,5,1,10,0], [0,1,1,1,1], @linear_bkgd, [0,0]);
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_1.m:110]: why are these disabled?

File: _test/test_multifit/test_multifit_horace_dnd_1.m [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            
+            assertTrue(equal_to_tol([wfit_single1,wfit_single2],wfit_single12),'fit_sqw workspaces not working');
+            assertTrue(equal_to_tol([fitpar_single1,fitpar_single2],fitpar_single12),'fit_sqw fitting not working')
+            
+%             % Test against saved or store to save later
+%             this=save_or_test_variables(this,wfit_single1,wfit_single2,wfit_single12);
+        end
+        
+        % ------------------------------------------------------------------------------------------------
+        function this=DISABLED_test_fit_test_fit_array_of_datasets_2(this)
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_1.m:141]: why is this disabled?

File: _test/test_multifit/test_multifit_horace_dnd_1.m [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            fitpar_single12(2).corr=[];
+            
+%             tol = this.tol;
+%             this.tol = -1;           
+%             this=save_or_test_variables(this,fitpar_single1,fitpar_single2,fitpar_single12);
+%             this.tol=tol;
+        end
+        
+        % ------------------------------------------------------------------------------------------------
+        function this=DISABLED_test_fit_array_of_datasets_3(this)
```
Comment (nickbattam-tessella) [inline _test/test_multifit/test_multifit_horace_dnd_1.m:167]: why is this disabled

File: _test/test_multifit/test_multifit_horace_dnd_1.m (hunk 6 above)
Comment (nickbattam-tessella) [reply _test/test_multifit/test_multifit_horace_dnd_1.m:141]: the name doesn't tell me what it's testing and how it's different to `..._array_of_datasets` or `..._array_of_datasets_3`

File: _test/test_multifit/test_multifit_horace_dnd_1.m (hunk 7 above)
Comment (nickbattam-tessella) [reply _test/test_multifit/test_multifit_horace_dnd_1.m:167]: the name doesn't tell me what it's testing and how it's different to `..._array_of_datasets` or `..._array_of_datasets_2`

**Review by nickbattam-tessella — COMMENTED** [review]
The `multifit_*.m` functions call mfclass with `sqw` as the `dataset_class` argument even when called from DnD; should this be `class(obj)`?

File: horace_core/sqw/@d1d/d1d.m [hunk 8]
```diff
@@ -30,5 +30,32 @@
...
+            %   >> obj = loadobj(S)
+            %
+            % Input:
+            % ------
+            %   S       An instance of this object or struct
+            %
+            % -------
+            % Output:
+            %   obj     An instance of this object
+            obj = d1d(S);
```
Comment (nickbattam-tessella) [inline horace_core/sqw/@d1d/d1d.m:45]: not sure this is useful: the object case is covered below (for one and array arg) and the struct case below that (one and array cases)

File: _test/test_multifit/test_multifit_horace_dnd_1.m (hunk 5 above)
Comment (cmarooney-stfc) [reply _test/test_multifit/test_multifit_horace_dnd_1.m:110]: @nickbattam-tessella Originally these tests were disabled as the loader had not been prepared. Now, having replaced fit_sqw (obsolete) with multifit_sqw, they fail as func_eval is not ported to d1d. Simple port would accept the conversion d1d->sqw->eval->d1d, but do we want to do that at this point?

File: _test/test_multifit/test_multifit_horace_dnd_2.m (hunk 1 above)
Comment (cmarooney-stfc) [reply _test/test_multifit/test_multifit_horace_dnd_2.m:1]: Undoubtedly. But this naming mirrors what is in the sqw equivalent .m files.

Replying to the observation that the multifit_*.m functions call mfclass with sqw as dataset class - the only effect that has (method: find "dataset_class" in all of Horace and Herbert) is that if the average property has been set for this multifit object, the using sqw will cause a more local average flag to be passed into the operation. Also noticing that while mfclass_horace_sqw_sqw does this with sqw, mfclass_Horace_sqw still does this with sqw_old. Guessing that the averaging may be a pixel data only process.

File: _test/test_multifit/test_multifit_horace_dnd_1.m (hunk 5 above)
Comment (oerc0122) [reply _test/test_multifit/test_multifit_horace_dnd_1.m:110]: Would prefer disabling by skipTest

File: horace_core/sqw/@d0d/d0d.m [hunk 9]
```diff
@@ -29,5 +29,32 @@
...
+            %   >> obj = loadobj(S)
+            %
+            % Input:
+            % ------
+            %   S       An instance of this object or struct
+            %
+            % -------
+            % Output:
+            %   obj     An instance of this object
+            obj = d0d(S);
```
Comment (oerc0122) [inline horace_core/sqw/@d0d/d0d.m:44]: I would move this after the `if` to keep it obvious that this has no effect on the if clause.

File: horace_core/sqw/@d0d/d0d.m [hunk 10]
```diff
@@ -29,5 +29,32 @@
...
+            if isa(S,'d0d')
+               obj = S;
+               return
+            end
+            if numel(S)>1
+               tmp = d0d();
+               obj = repmat(tmp, size(S));
+               for i = 1:numel(S)
```
Comment (oerc0122) [inline horace_core/sqw/@d0d/d0d.m:52]: Would this be more obvious as an `arrayfun`? Also removes the need to preallocate

**Pull request closed without merging**